src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -0,0 +1,571 @@
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Top level IsStream module that can use all other lower level IsStream
+-- modules.
+
+module Streamly.Internal.Data.Stream.IsStream.Top
+    (
+    -- * Transformation
+    -- ** Sampling
+    -- | Value agnostic filtering.
+      sampleFromThen
+    , sampleIntervalStart
+    , sampleIntervalEnd
+    , sampleBurstStart
+    , sampleBurstEnd
+
+    -- * Nesting
+    -- ** Set like operations
+    -- | These are not exactly set operations because streams are not
+    -- necessarily sets, they may have duplicated elements.
+    , intersectBy
+    , mergeIntersectBy
+    , differenceBy
+    , mergeDifferenceBy
+    , unionBy
+    , mergeUnionBy
+
+    -- ** Join operations
+    , crossJoin
+    , innerJoin
+    , mergeInnerJoin
+    , hashInnerJoin
+    , leftJoin
+    , mergeLeftJoin
+    , hashLeftJoin
+    , outerJoin
+    , mergeOuterJoin
+    , hashOuterJoin
+    )
+where
+
+#include "inline.hs"
+
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict (get, put)
+-- import Data.Hashable (Hashable)
+import Data.IORef (newIORef, readIORef, modifyIORef')
+import Data.Kind (Type)
+#if !(MIN_VERSION_base(4,11,0))
+import Data.Semigroup (Semigroup(..))
+#endif
+import Streamly.Internal.Data.SVar (MonadAsync)
+import Streamly.Internal.Data.Stream.IsStream.Common (concatM)
+import Streamly.Internal.Data.Stream.Prelude (foldl', fromList)
+import Streamly.Internal.Data.Stream.Serial (SerialT)
+import Streamly.Internal.Data.Stream.StreamK (IsStream)
+import Streamly.Internal.Data.Time.Units (NanoSecond64(..), toRelTime64)
+
+import qualified Data.List as List
+import qualified Streamly.Internal.Data.Array as Array
+import qualified Streamly.Internal.Data.Fold as Fold
+import qualified Streamly.Internal.Data.Stream.IsStream.Lift as Stream
+import qualified Streamly.Internal.Data.Stream.IsStream.Eliminate as Stream
+import qualified Streamly.Internal.Data.Stream.IsStream.Nesting as Stream
+import qualified Streamly.Internal.Data.Stream.IsStream.Transform as Stream
+import qualified Streamly.Internal.Data.Stream.StreamK as StreamK
+
+import Prelude hiding (filter, zipWith, concatMap, concat)
+
+------------------------------------------------------------------------------
+-- Sampling
+------------------------------------------------------------------------------
+
+-- | @sampleFromthen offset stride@ samples the element at @offset@ index and
+-- then every element at strides of @stride@.
+--
+-- @
+-- >>> Stream.toList $ Stream.sampleFromThen 3 2 $ S.enumerateFromTo 0 10
+-- > [1,4,7]
+-- @
+--
+-- /Internal/
+--
+{-# INLINE sampleFromThen #-}
+sampleFromThen :: (IsStream t, Monad m, Functor (t m)) =>
+    Int -> Int -> t m a -> t m a
+sampleFromThen size offset =
+    Stream.with Stream.indexed Stream.filter
+        (\(i, _) -> (i + offset) `mod` size == 0)
```

Comment:
Shouldn't this be `(i - offset) `mod` size`?

Reply:
Both are correct (unless `i + offset` overflows and `i - offset` doesn't). But the condition has to be changed like this - `(\(i, _) -> i >= offset && (i - offset) `mod` size == 0)`

Reply:
How is it `i + offset`?
If offset is 5, then first element in the stream should be at index 5 (we skip 5 elements)
`(i - 5) "mod" stride == 0` as `(i - 5) == 0`
`(i + 5) "mod" stride` is not guarenteed to be `0`

Reply:
Yes, you are right.